### PR TITLE
[cmake] Respect CLANG_LINK_CLANG_DYLIB for objlibs

### PR DESCRIPTION
**Pull request by nikic**
add_clang_library() will create a library clangFoo that depends on objlib obj.clangFoo. Then clang_target_link_libraries() will add a clang-cpp dependency to clangFoo, but obj.clangFoo will instead get dependencies on the individual clangXYZ libraries, instead of the shared object. The final outcome is that we will link both the static libraries and the shared object, leading to an increase in binary size and link times (when using LTO).

Make sure we use the same logic for the obj and non-obj libs.

I believe this regressed with the addition of the llvm-driver functionality in https://github.com/llvm/llvm-project/commit/f06abbb393800b0d466c88e283c06f75561c432c.

### CODE REVIEW

**Comment by nikic** [comment]
ping

File: clang/cmake/modules/AddClang.cmake [hunk 1]
```diff
@@ -194,18 +194,20 @@ macro(add_clang_symlink name dest)
...
-
   get_property(LLVM_DRIVER_TOOLS GLOBAL PROPERTY LLVM_DRIVER_TOOLS)
   if(LLVM_TOOL_LLVM_DRIVER_BUILD AND ${target} IN_LIST LLVM_DRIVER_TOOLS)
     set(target llvm-driver)
   endif()
 
   if (CLANG_LINK_CLANG_DYLIB)
     target_link_libraries(${target} ${type} clang-cpp)
+    if (TARGET obj.${target})
+      target_link_libraries(obj.${target} clang-cpp)
```
Comment (llvm-beanz) [inline clang/cmake/modules/AddClang.cmake:205]: Should we specify this as an `INTERFACE` link dependency?

```suggestion
      target_link_libraries(obj.${target} INTERFACE clang-cpp)
```

File: clang/cmake/modules/AddClang.cmake [hunk 2]
```diff
@@ -194,18 +194,20 @@ macro(add_clang_symlink name dest)
...
+    endif()
   else()
     target_link_libraries(${target} ${type} ${ARGN})
+    if (TARGET obj.${target})
+      target_link_libraries(obj.${target} ${ARGN})
```
Comment (llvm-beanz) [inline clang/cmake/modules/AddClang.cmake:210]:
```suggestion
      target_link_libraries(obj.${target} INTERFACE ${ARGN})
```

**Review by llvm-beanz — APPROVED** [review]
I think dependencies for object libraries are implicitly `INTERAFCE`, but it might be nice to make that explicit. I know CMake complains if you mix and match implicit and explicit specification of linkage types.

Otherwise LGTM.

**Comment by nikic** [comment]
I suspect I may have misdiagnosed the issue, and the actual problem may be just https://github.com/llvm/llvm-project/pull/94588. I'll double check whether this change is necessary before landing.

**Pull request closed without merging**